### PR TITLE
Implement epochs

### PR DESCRIPTION
**Pull request by vladshablinsky**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Issue: https://github.com/Homebrew/legacy-homebrew/issues/41563

The idea is to add `epoch` to DSL and reimplement `outdated_versions`-related things to be able to change version schemes, thus epochs are quite similar to revisions, but we store them in tabs.

CC @xu-cheng

### CODE REVIEW

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -1013,9 +1018,10 @@ def _outdated_versions(options = {})
 
     installed_kegs.each do |keg|
       version = keg.version
+      tab = Tab.for_keg(keg)
       all_versions << version
 
-      return [] if pkg_version <= version && !version.head?
+      return [] if pkg_version <= version && epoch == tab.epoch && !version.head?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: Can you split this into multiple statements/lines somehow? Quite a bit of logic being chained here; perhaps having 2-3 separate variables would help.

File: Library/Homebrew/formula.rb [hunk 2]
```diff
@@ -1695,6 +1702,15 @@ class << self
     # <pre>revision 1</pre>
     attr_rw :revision
 
+    # @!attribute [w] revision
+    # Used for creating new Homebrew versions shemes. For example, if we want
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: `shemes` -> `schemes`

File: Library/Homebrew/formula.rb [hunk 3]
```diff
@@ -1695,6 +1702,15 @@ class << self
...
+    # to change version scheme from one to another, then we may need to update
+    # `epoch` of this {Formula} to be able to use new version scheme.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: Might want to note explicitly the e.g. moving from 20151020 scheme to 1.0.0 and note that, without this, the prior scheme will always equate to a higher version.

**Comment by MikeMcQuaid** [comment]
Great work on this, @vladshablinsky. Could you make a PR to homebrew-core with an example that we could use to test (and merge after this is done) and some commands to run to test this?

**Comment by MikeMcQuaid** [comment]
CC @DomT4 too who will be interested in this.

File: Library/Homebrew/formula.rb (hunk 1 above)
Comment (vladshablinsky) [inline Library/Homebrew/formula.rb]: I can come up with this:

```ruby
      if !version.head?
        return [] if epoch == tab.epoch && pkg_version <= version
      end
```

Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: How about:
```ruby
outdated_version = pkg_version <= version
equal_epoch = epoch == tab.epoch
not_head_version = !version.head?
return [] if outdated_version && equal_epoch && not_head_version

```

Comment (vladshablinsky) [inline Library/Homebrew/formula.rb]: `outdated_version` is actually `not_outdated_version`.

I don't really like `not` prefixes, but If you're OK with them I make the change. Also, `equal_epoch` can be done `same_epoch`.

Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: `latest_version`, `same_epoch`, `not_head` perhaps (as I can't think of a way of decribing "not head" without "not").

Comment (vladshablinsky) [inline Library/Homebrew/formula.rb]: :+1: will make the change.

**Comment by DomT4** [comment]
Thanks for working on this @vladshablinsky ❤️. `osh` is a formula in `homebrew/core` that is known to need epochs to deal with a version change shift, so if you want to play with one that'd be a solid candidate possibly.

Will play with this later tonight locally a bit; currently working on GPG stuff, which is being "fun".

**Comment by MikeMcQuaid** [comment]
I tested this with https://github.com/Homebrew/homebrew-core/issues/3900 and it worked great. I installed the `20160726` version, applied https://github.com/Homebrew/homebrew-core/issues/3900 as a patch, removed the custom version stuff in there in favour of a `epoch 1`, verified that `brew outdated` and `brew upgrade` showed it as outdated and upgraded to `1.0.0`, uninstalled the `20160726` version manually, reverted https://github.com/Homebrew/homebrew-core/issues/3900, added `epoch 2`, verified that `brew outdated` and `brew upgrade` showed it as outdated and upgraded to `20160726`.

As far as I'm concerned this is good to 🚢. The only wording I might want to adjust is `epoch` to something more obvious like `version_scheme`, `version_scheme_revision`, `version_scheme_changed`, `version_scheme_version` etc. The overhead of making this more readable is very low and makes it much more obvious what it's doing, conceptually.

**Comment by MikeMcQuaid** [comment]
Also: I want to get this merged sooner rather than later so if @DomT4 agrees on one of those names let's try and get this shipped this week.

**Comment by vladshablinsky** [comment]
Will change it to `version_scheme` for now, I like this option the most.
Also, I'd like to say that this way of handling epochs was suggested by @xu-cheng, so thanks to him as well.

File: Library/Homebrew/formula.rb [hunk 4]
```diff
@@ -1013,9 +1018,13 @@ def _outdated_versions(options = {})
...
     installed_kegs.each do |keg|
       version = keg.version
+      tab = Tab.for_keg(keg)
       all_versions << version
 
-      return [] if pkg_version <= version && !version.head?
+      latest_version = pkg_version <= version
+      same_version_scheme = version_scheme == tab.version_scheme
+      not_head = !version.head?
+      return [] if latest_version && same_version_scheme && not_head
```
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: I think this should be:
```ruby
latest_version = pkg_version <= version
latest_version_scheme = version_scheme <= tab.version_scheme
not_head = !version.head?
return [] if (latest_version || latest_version_scheme) && not_head

Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: Let's pull it out to:
```ruby
latest_version = pkg_version <= version
latest_version_scheme = version_scheme <= tab.version_scheme
latest_version_or_scheme = latest_version || latest_version_scheme
not_head = !version.head?
return [] if latest_version_or_scheme && not_head
```

(to simplify humans parsing the binary logic)

Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: How about:
```ruby
 installed_kegs.each do |keg|
  version = keg.version
  tab = Tab.for_keg(keg)
  all_versions << version
  next if version.head?
  next if version_scheme > tab.version_scheme
  next if pkg_version > version
  return []
end
```

Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: Even better. Could skip creating the `tab` until after `version.head?` when we know it's needed, too.

**Comment by DomT4** [comment]
I like `epoch` because the concept should be immediately _slightly_ more familiar to anyone who has encountered Debian's packaging before, but if the cost of getting this done is losing `epoch` it's not the worst thing in the world.

I can live with `version_scheme`. It feels a bit strange to write but I suspect that's down to my usage of `epoch` for the last year every time this discussion has come up.

I think it's worth adding `audit` code similar to our revision check to ensure that `epochs` or whatever it comes to be called are never removed from formulae, because obviously once it's there if a contributor removes it that'll cause upgrade problems and that may not be immediately obvious to every contributor, especially since we _do_ remove `revision`s on version updates.

**Comment by MikeMcQuaid** [comment]
> I like epoch because the concept should be immediately slightly more familiar to anyone who has encountered Debian's packaging before

FWIW I'd done Debian packaging before and never heard epoch used this way so I think this is catering for a very small group of people.



> I think it's worth adding audit code similar to our revision check to ensure that epochs or whatever it comes to be called are never removed from formulae, because obviously once it's there if a contributor removes it that'll cause upgrade problems and that may not be immediately obvious to every contributor, especially since we do remove revisions on version updates.

Agreed (but in a follow-up PR).

**Comment by DomT4** [comment]
> Agreed (but in a follow-up PR).

Doesn't need to be a blocker here but would like it chased up on before we start using this in `homebrew/core`. Documentation in the Cookbook would be a 👍 as well, although obviously there's broader changes to come there if we're going to be finally renaming revisions et al.

File: Library/Homebrew/formula.rb (hunk 4 above)
Comment (xu-cheng) [inline Library/Homebrew/formula.rb]: Actually, I just realize I give an incorrect suggestion in the first place. The correct way is:
```ruby
installed_kegs.each do |keg|
  version = keg.version
  tab = Tab.for_keg(keg)
  all_versions << version
  next if version.head?
  next if version_scheme > tab.version_scheme
  next if version_scheme == tab.version_scheme && pkg_version > version
  return []
end
```

Noted that the correct way to compare two version is:
```ruby
(version_scheme <=> other.version_scheme).nonzero? || pkg_version <=> other.pkg_version
```

Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: Would be good to make `next if version_scheme == tab.version_scheme && pkg_version > version` into two variables which are `&&` to again make the boolean logic better.

Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb]: And, again, moving `Tab.for_keg(keg)` after `version.head?` so it's not read unnecessariy.

**Comment by MikeMcQuaid** [comment]
Tested this again locally and it worked great. Nice work @vladshablinsky 👏

**Comment by MikeMcQuaid** [comment]
> Doesn't need to be a blocker here but would like it chased up on before we start using this in homebrew/core. Documentation in the Cookbook would be a 👍 as well, although obviously there's broader changes to come there if we're going to be finally renaming revisions et al.

Done in https://github.com/Homebrew/brew/pull/743.

**Comment by dunn** [comment]
I saw this too late, but IMO `version_schema` is a misleading name since it suggests a _schema_, whereas what we've got is something more like an iteration or, well, an epoch.

**Comment by MikeMcQuaid** [comment]
@dunn It's a migration between versioning schemes e.g. `0` is one versioning scheme, `1` is another etc.

**Comment by dunn** [comment]
Oh I know, but to me it suggests the DSL is for _specifying_ the scheme, rather than the versioning scheme ... version 😸

**Comment by ilovezfs** [comment]
Are updates to taps occurring before updates to the main brew code?
```
Josephs-MacBook-Pro:~ joe$ brew update
Updated Homebrew from db2e9b8 to 5c7c9de.
Error: undefined method `version_scheme' for #<Class:0x007fb2ab9d91e8>
/usr/local/Library/Taps/homebrew/homebrew-core/Formula/pyenv.rb:6:in `<class:Pyenv>'
/usr/local/Library/Taps/homebrew/homebrew-core/Formula/pyenv.rb:1:in `load_formula'
/usr/local/Library/Homebrew/formulary.rb:25:in `module_eval'
/usr/local/Library/Homebrew/formulary.rb:25:in `load_formula'
/usr/local/Library/Homebrew/formulary.rb:42:in `load_formula_from_path'
/usr/local/Library/Homebrew/formulary.rb:91:in `load_file'
/usr/local/Library/Homebrew/formulary.rb:82:in `klass'
/usr/local/Library/Homebrew/formulary.rb:78:in `get_formula'
/usr/local/Library/Homebrew/formulary.rb:215:in `factory'
/usr/local/Library/Homebrew/cmd/update-report.rb:212:in `block in report'
/usr/local/Library/Homebrew/cmd/update-report.rb:190:in `each_line'
/usr/local/Library/Homebrew/cmd/update-report.rb:190:in `report'
/usr/local/Library/Homebrew/cmd/update-report.rb:370:in `add'
/usr/local/Library/Homebrew/cmd/update-report.rb:62:in `block in update_report'
/usr/local/Library/Homebrew/tap.rb:478:in `block (2 levels) in each'
/usr/local/Library/Homebrew/tap.rb:477:in `each'
/usr/local/Library/Homebrew/tap.rb:477:in `block in each'
/usr/local/Library/Homebrew/tap.rb:476:in `each'
/usr/local/Library/Homebrew/tap.rb:476:in `each'
/usr/local/Library/Homebrew/cmd/update-report.rb:52:in `update_report'
/usr/local/Library/Homebrew/brew.rb:87:in `<main>'
Updated 5 taps (caskroom/cask, homebrew/bundle, homebrew/core, homebrew/php, homebrew/versions).
==> New Formulae
homebrew/php/php70-gearman        homebrew/php/php71-snmp           homebrew/php/php71-timecop      
homebrew/php/php71-oauth          homebrew/php/php71-stats          homebrew/php/php71-xxtea        
homebrew/php/php71-pcntl          homebrew/php/php71-swoole         homebrew/versions/go16          
homebrew/php/php71-pspell         homebrew/php/php71-tidy         
==> Updated Formulae
homebrew/php/phpmyadmin                            pyenv                                            
Josephs-MacBook-Pro:~ joe$ 
```

**Comment by MikeMcQuaid** [comment]
@ilovezfs They are not but did you have a non-master `Homebrew/brew` branch checked out?

**Comment by ilovezfs** [comment]
ah nevermind I think I know what happened.

**Comment by MikeMcQuaid** [comment]
Also that exception is only printed for developers.

**Comment by ilovezfs** [comment]
@MikeMcQuaid yeah, that particular tree is on non-master since it has my version of bump-formula-pr.

**Comment by MikeMcQuaid** [comment]
> my version of bump-formula-pr.

PR requested 😉

**Pull request merged**